uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js

```diff
@@ -0,0 +1,492 @@
+
+        if (!dateParam) {
+          deferred.reject([]);
+        }
+
+        var date = convertDateFormatToServer(dateParam);
+        // Make a copy of the list
+        var listToReturn = $ctrl.leaveRequestDayTypes.slice(0);
+
+        PublicHoliday.isPublicHoliday(date).then(function (result) {
+          if (result) {
+
+            listToReturn = listToReturn.filter(function (item) {
+              return item.name === 'public_holiday';
+            });
+          } else {
+            //if not found the calender methods throw exceptions
+            var foundInCalendar = false;
+
+            try {
+              if ($ctrl.calendar.isNonWorkingDay(date)) {
+                // Only 'Non Working Day' option
+                listToReturn = listToReturn.filter(function (item) {
+                  return item.name === 'non_working_day';
+                });
+                foundInCalendar = true;
+              } else if ($ctrl.calendar.isWeekend(date)) {
+                // Only 'Weekend' option
+                listToReturn = listToReturn.filter(function (item) {
+                  return item.name === 'weekend';
+                });
+                foundInCalendar = true;
+              }
+            } catch (e) {
+              //empty catch to catch exceptions
+            } finally {
+              if (!foundInCalendar) {
+                // 'All day', '1/2 AM', and '1/2 PM' options
+                listToReturn = listToReturn.filter(function (item) {
+                  return item.name === 'all_day' || item.name === 'half_day_am' || item.name === 'half_day_pm';
+                });
+              }
+            }
+          }
+
+          if (isFrom) {
+            $ctrl.leaveRequestFromDayTypes = listToReturn;
+            $ctrl.uiOptions.selectedFromType = $ctrl.leaveRequestFromDayTypes[0];
+            $ctrl.leaveRequest.from_date_type = $ctrl.uiOptions.selectedFromType.name;
+          } else {
+            $ctrl.leaveRequestToDayTypes = listToReturn;
+            $ctrl.uiOptions.selectedToType = $ctrl.leaveRequestToDayTypes[0];
+            $ctrl.leaveRequest.to_date_type = $ctrl.uiOptions.selectedToType.name;
+          }
+
+          deferred.resolve(listToReturn);
+        });
+
+        return deferred.promise;
+      }
+
+      /**
+       * helper function to reset pagination for balance breakdow
+       *
+       **/
+      function rePaginate() {
+        $ctrl.pagination.totalItems = $ctrl.balance.change.breakdown.length;
+        $ctrl.pagination.filteredbreakdown = $ctrl.balance.change.breakdown;
+        $ctrl.pagination.pageChanged();
+      }
+
+      /**
+       * Pick a specific value out of a collection
+       *
+       * @param {array} the option group collection key
+       * @param {string} key - The sub-collection key
+       * @param {string} value - The sub-collection key's value to match
+       * @return {object}
+       */
+      function getSpecificValueFromCollection(collection, key, value) {
```

Comment:
You are using this only once, is this really necessary? This looks awfully generic for a one-time specific purpose use